### PR TITLE
Support interfaces whose symbols are already declared (towards #1618)

### PR DESCRIPTION
**Pull request by rupertford**
The Met Office have algorithm code that looks like the following:

```
module test_mod
  type, public :: test_type
  end type
  interface test_type
     module procedure constructor
  end interface
contains
  function constructor()
  end function
end module
```

This causes problems for the PSyIR as the PSyIR assumes that an interface symbol has not been declared. This PR provides a fix to this problem. This fix is not a long term solution and should be replaced when interfaces are better supported in the PSyIR.

### CODE REVIEW

**Comment by rupertford** [comment]
One for @sergisiso I think as I suspect he will have an opinion on it.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@rupertford I am fine with this since it appears in LFRic algorithms and putting the whole module in a CodeBlock would not be an alternative. See a couple of inline questions.

Also, it may be good to add a note/section in https://psyclone-dev.readthedocs.io/en/latest/psyir_symbols.html saying that interfaces are RoutineSymbols with UnknownFortranType and they have these "unused" names when there is name clashes with a type.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 1]
```diff
@@ -2088,9 +2088,20 @@ def process_declarations(self, parent, nodes, arg_list,
...
+                        RoutineSymbol(
+                            name, UnknownFortranType(str(node).lower()),
+                            visibility=vis))
+                except KeyError:
+                    # This symbol has already been declared. However
+                    # we still want to output the interface so we
+                    # store it in the PSyIR as an UnkownFortranType
+                    # with a different name.
+                    parent.symbol_table.new_symbol(
+                        root_name=name, symbol_type=RoutineSymbol,
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/frontend/fparser2.py:2102]: Good idea about the name clash. I've gone with `f_psyclone_internal_{name}"` i.e. added an additional underscore at the start as Fortran symbols won't have this.

[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/frontend/fparser2.py:2102]: Yes, I think the same thing could be done for interfaces with operators if we want a quick fix. However, might it not be better for us to look to fix interfaces properly? I would suggest a new type of symbol - an InterfaceSymbol presumably, which could be named (with a RoutineSymbol?), or reference a type, or overload an operator and would contain a list of references to routine symbols. I guess this needs discussion. I'll bring it up at the next telco.

**Comment by rupertford** [comment]
Ready for next review from @sergisiso

**Review by sergisiso — APPROVED** [review]
@rupertford Thanks for adding the documentation. All comments have been addressed, there are TODOs remaining but this is a towards PR. Pylint and pycodestyle are clean. All tests pass with compilation. I tested that gravity_wave and gungho still work (although I don't know if the algorithm parsing is already affected by this change or will be in the future). This is ready to merge.

**Pull request merged**